pow/src/main/java/tech/pegasys/artemis/DepositRequestManager.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static tech.pegasys.artemis.util.config.Constants.ETH1_FOLLOW_DISTANCE;
+
+import com.google.common.primitives.UnsignedLong;
+import io.reactivex.disposables.Disposable;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes32;
+import org.web3j.protocol.Web3j;
+import org.web3j.protocol.core.DefaultBlockParameter;
+import org.web3j.protocol.core.methods.response.EthBlock;
+import tech.pegasys.artemis.pow.contract.DepositContract;
+import tech.pegasys.artemis.pow.event.Deposit;
+import tech.pegasys.artemis.pow.event.DepositsFromBlockEvent;
+import tech.pegasys.artemis.util.async.SafeFuture;
+
+public class DepositRequestManager {
+
+  private final Web3j web3j;
+  private final Consumer<DepositsFromBlockEvent> depositsFromBlockEventConsumer;
+  private final DepositContract depositContract;
+  private static final Logger LOG = LogManager.getLogger();
+
+  private boolean active = false;
+  private boolean requestQueued = false;
+  private BigInteger latestCanonicalBlockNumber;
+  private BigInteger latestSuccessfullyQueriedBlockNumber = BigInteger.ZERO;
+  private Disposable newBlockSubscription;
+
+  public DepositRequestManager(
+      Web3j web3j,
+      Consumer<DepositsFromBlockEvent> depositsFromBlockEventConsumer,
+      DepositContract depositContract) {
+    this.web3j = web3j;
+    this.depositsFromBlockEventConsumer = depositsFromBlockEventConsumer;
+    this.depositContract = depositContract;
+  }
+
+  public void start() {
+    newBlockSubscription =
+        web3j
+            .blockFlowable(false)
+            .map(EthBlock::getBlock)
+            .map(EthBlock.Block::getNumber)
+            .map(number -> number.subtract(ETH1_FOLLOW_DISTANCE.bigIntegerValue()))
+            .subscribe(
+                this::updateLatestCanonicalBlockNumber,
+                err -> {
+                  LOG.warn("New block subscription failed, retrying.", err);
+                  start();
```

Comment:
Probably should add a `newBlockSubscription.dispose()` prior to this line just for good measure.

Reply:
Done.